store/tikv/lock_resolver.go

```diff
@@ -342,7 +351,9 @@ func (lr *LockResolver) resolveLocks(bo *Backoffer, callerStartTS uint64, locks
 				cleanTxns[l.TxnID] = cleanRegions
 			}
 
-			if l.LockType == kvrpcpb.Op_PessimisticLock {
+			if status.useAsyncCommit {
```

Comment:
I found `status.ttl` is never 0 in this case. There are two choices:
1. backoff in TiDB until the TTL has elapsed.
2. handle async-commit locks specially in TiKV. When the lock has expired, don't roll back it and return ttl=0.

Reply:
I prefer 2 because it's simple.

Reply:
Why isn't the TTL 0 when the lock has expired?

Reply:
TiKV never thinks it's expired because currentTS is 0.

Reply:
https://github.com/tikv/tikv/blob/46ab9542908bc43ab04917e0c5d827b459874de1/src/storage/txn/commands/check_txn_status.rs#L121

We can change the TTL=0 here if it's an async-commit txn: https://github.com/pingcap/tidb/pull/18467/files#diff-39b30a313dde4d126e56f1828b6dabdcR473.

Reply:
I don't think we should do async commit recovery regardless of the actual TTL. It should do the recovery until backoff.

Reply:
Of course, check with `Oracle.IsExpired()`

Reply:
Hmm, I assumed we already did 2, I think that is what should happen

Reply:
@sticnarf What do you think?

Reply:
Do you mean we pass the correct `currentTS` to TiKV so TiKV knows if the lock is actually expired?

Reply:
We need to keep the possibility for fallback. If `currentTS` is passed to TiKV and TiKV handles async commit locks specially, then how to do `CheckTxnStatus` if we want to fallback?

Reply:
I prefer to still keep the current way but change `if status.ttl == 0` to `if status.ttl == 0 || oracle.expired(start_ts, ttl)`. wdyt?

Reply:
I'm fine with either.

Reply:
At first, I think we could wrap some logic in the `getTxnStatusFromLock` function or call another function like this, which will make the transaction status certain for `async commit protocol`. The returned status is used to judge we should resolve the lock or report key is locked. Currently `resolveLockAsync` has combined them together and we need to be cautious about the original conditions and presumes.